libs/pdi-scanner/src/rust/client.rs

```diff
@@ -126,57 +79,81 @@ impl<T> Client<T> {
                 "failed to send packet to scanner (host to scanner channel closed)",
             )))
         })?;
-        let ack_id = self.host_to_scanner_ack_rx.recv().map_err(|_| {
-            Error::Usb(UsbError::Nusb(nusb::Error::new(
+        let Some(ack_id) = self.host_to_scanner_ack_rx.recv().await else {
+            return Err(Error::Usb(UsbError::Nusb(nusb::Error::new(
                 io::ErrorKind::ConnectionAborted,
                 "failed to receive ack from scanner (host to scanner ack channel closed)",
-            )))
-        })?;
+            ))));
+        };
         assert_eq!(id, ack_id);
         Ok(())
     }
 
+    // There should only be one command/response occurring at a time, so before
+    // we send a command that expects a response, we should clear out any old
+    // responses. This ensures we don't get an outdated response (e.g. a previous scanner
+    // status that didn't get received when its command was sent).
+    //
+    // This accounts for the fact that there are cases where the scanner will
+    // delay sending a response to a command (e.g. when the "eject pause"
+    // feature pauses the scanner, it will queue commands received during that
+    // time).
+    fn clear_unhandled_solicited_packets(&mut self) {
+        let mut unhandled_packets = VecDeque::with_capacity(self.unhandled_packets.len());
+        std::mem::swap(&mut unhandled_packets, &mut self.unhandled_packets);
+
+        let (unsolicited_packets, solicited_packets): (Vec<_>, Vec<_>) = unhandled_packets
+            .into_iter()
+            .partition(|packet| packet.message_type().is_unsolicited());
```

Comment:
Here is the updated code where we filter packets. It’s done a little differently than the previous version in that we only do the filter/partition once now and use `std::mem::swap` to pull the data out of `self`, then put it back below with `VecDeque::extend`.